google-home-community.groovy

```diff
@@ -257,297 +297,331 @@ private hubVersionLessThan(versionString) {
         }
     }
 }
 
 @SuppressWarnings('MethodSize')
-def deviceTypePreferences(deviceType) {
-    state.remove("currentlyEditingDeviceTrait")
+def deviceTypePreferences(Map ideviceType=null) {
+    state.remove('currentlyEditingDeviceTrait')
+    Map<String,Object> deviceType; deviceType= ideviceType
     if (deviceType == null) {
-        deviceType = deviceTypeFromSettings(state.currentlyEditingDeviceType)
+        deviceType = deviceTypeFromSettings(gtStStr('currentlyEditingDeviceType'))
     }
 
-    if (settings.deviceTraitToAdd != null) {
-        def toAdd = settings.deviceTraitToAdd
-        app.removeSetting("deviceTraitToAdd")
-        def traitName = "${deviceType.name}.traits.${toAdd}"
-        addTraitToDeviceTypeState(deviceType.name, toAdd)
-        return deviceTraitPreferences([name: traitName])
-    }
+    if(deviceType!=null){
+        String toAdd = gtSetStr('deviceTraitToAdd')
+        if (toAdd != sNL) {
+            app.removeSetting('deviceTraitToAdd')
+            String traitName = "${deviceType.name}.traits.${toAdd}"
+            addTraitToDeviceTypeState((String)deviceType.name, toAdd)
+            return deviceTraitPreferences([(sNM): traitName])
+        }
 
-    if (state.pinToDelete) {
-        deleteDeviceTypePin(deviceType, state.pinToDelete)
-        state.remove("pinToDelete")
+        String pinDel= gtStStr('pinToDelete')
+        if (pinDel) {
+            deleteDeviceTypePin(deviceType, pinDel)
+            state.remove('pinToDelete')
+        }
+    } else {
+        app.removeSetting('deviceTraitToAdd')
+        state.remove('pinToDelete')
+        (iZ..<state.nextDeviceTypeIndex).each { i ->
+            deviceType = deviceTypeFromSettings("deviceTypes.${i}",true)
+            //LOGGER.debug("deviceTypePreferences() $i checking $deviceType")
+            if (deviceType != null && !deviceType.traits) {
+                deleteDeviceType(deviceType)
+            }
+        }
+        deviceType=null
     }
 
-    return dynamicPage(name: "deviceTypePreferences", title: "Device Type Definition", nextPage: "mainPreferences") {
-        def devicePropertyName = deviceType != null ? deviceType.name : "deviceTypes.${state.nextDeviceTypeIndex++}"
+    return dynamicPage((sNM): "deviceTypePreferences", (sTIT): "Device Type Definition", nextPage: "mainPreferences") {
+        String devicePropertyName = deviceType != null ? deviceType.name : "deviceTypes.${state.nextDeviceTypeIndex++}"
         state.currentlyEditingDeviceType = devicePropertyName
         section {
             input(
-                name: "${devicePropertyName}.display",
-                title: "Device type name",
-                type: "text",
-                required: true
+                (sNM): "${devicePropertyName}.display",
+                (sTIT): "Device type name",
+                (sTYPE): sTEXT,
+                (sREQ): (!!gtSetStr("${devicePropertyName}.type") || !!gtSetStr("${devicePropertyName}.googleDeviceType"))
```

Comment:
Is the goal here to allow a device type to be deleted before it's fully filled out?